packages/@manta-style/plugin-system/docs/MockPlugin.md

```diff
@@ -4,7 +4,115 @@
 - Package name for a mock plugin should be `@manta-style/plugin-mock-*`
 
 ## Node type supported
+
 - ArrayType
 - StringType
 - NumberType
-- BooleanType
\ No newline at end of file
+- BooleanType
+
+## Plugin Interface
+
+```typescript
+type SupportedMockNodeType =
+  | 'ArrayType'
+  | 'StringType'
+  | 'NumberType'
+  | 'BooleanType';
+
+type Plugin = {
+  name: string;
+  mock: {
+    [key in SupportedMockNodeType]?: (
+      type: Type,
+      annotations: Annotation[],
+    ) => any
+  };
+};
+```
+
+# Builder Plugin
+
+- Builder plugin is responsible for compiling the config file into JavaScript that utilizes `@manta-style/runtime`
+- Package name for a mock plugin should be `@manta-style/plugin-builder-*`
+- Current language support
+
+  1. TypeScript (`@manta-style/plugin-builder-typescript`)
+  2. Flowtype (WIP, `@manta-style/plugin-builder-flow`)
+  3. GraphQL (Under consideration, `@manta-style/plugin-builder-graphql`)
+
+## Interface
+
+```typescript
+interface BuilderPlugin {
+  name: string;
+  supportedExtensions: string[];
+  build(configFilePath): Promise<Result>;
+}
+```
+
+# CLI Plugin
+
+- CLI Plugin is responsible for adding interfaces to the command line tool `ms`.
+- Package name for a CLI plugin should be `@manta-style/plugin-cli-*`
+- Current CLI plugin support
+  1. KOA REST Server (`@manta-style/plugin-cli-koa-rest`)
+  2. Boilerplate Code Generator (Under consideration, `@manta-style/plugin-cli-boilerplate-generator`)
+  3. Apollo GraphQL Server (Under consideration, `@manta-style/plugin-cli-apollo-graphql`)
+
+## Plugin Interface (TBD)
+
+```typescript
+interface CLIPlugin {
+  name: string;
+  launchCommands: Array<{ option: string; description: string }>;
+  activationKeys: Array<{ key: string; description: string }>;
+  launch(configs: any)?: Promise<void>;
+  didReceiveRequest()?: Promise<any>;
+  willSendResponse()?: Promise<any>;
+  didActivate(key: string)?: Promise<Array<{ key: string; description: string }>>;
+  willDestroy()?: Promise<void>;
+};
+```
```

Comment:
do we really need cli as plugins?
if we extract out MantaStyle as an API, we can write any cli for it right?
probably if you want, we can have a base cli, and have `@manta-style/cli-koa-rest`, `@manta-style/cli-boilerplate-generator`, etc extends it. if you want to abstract out the `commands` and `activationKeys`

Reply:
i feel some CLI features could compose together, rather than run several different command line tools.

Reply:
for example?
cause if for Snapshot, what i have in mind is like
```
const MantaStyleWithSnapshot = withSnapshot(MantaStyle)
```
prob the plugin is not for MantaStyle, but a plugin system for cli.

Reply:
for example, snapshot plugin can work on both `cli-koa-rest` and `cli-apollo-snapshot`. i think `withSnapshot` is also a good idea.